### PR TITLE
Use BlockingQueue.drainTo() in Journal thread

### PR DESCRIPTION
**Pull request by merlimat**
### Motivation

Use `BlockingQueue.drainTo()` in the Journal thread to minimize the contention between this thread and the threads passing entries to the journal. 

The drain method in the `ArrayBlockingQueue` method will use the mutex only once and take out all the items in one shot.

The logic in the journal dequeuing is already quite complicated. This change aims at not changing the whole logic, rather doing the minimal intervention.

### CODE REVIEW

**Review by zymap — APPROVED** [review]

**Comment by zymap** [comment]
rerun the failure tests

```
Error:  Failures: 
Error:  org.apache.bookkeeper.bookie.BookieJournalForceTest.testAckBeforeSyncWithJournalBufferedEntriesThreshold
Error:    Run 1: BookieJournalForceTest.testAckBeforeSyncWithJournalBufferedEntriesThreshold:251
Error:    Run 2: BookieJournalForceTest.testAckBeforeSyncWithJournalBufferedEntriesThreshold:251
Error:    Run 3: BookieJournalForceTest.testAckBeforeSyncWithJournalBufferedEntriesThreshold:251
[INFO] 
Warning:  Flakes: 
Warning:  org.apache.bookkeeper.bookie.StateManagerTest.testReadOnlyBookieTransitions
Error:    Run 1: StateManagerTest.testReadOnlyBookieTransitions:173->Object.wait:-2 » TestTimedOut
[INFO]   Run 2: PASS
[INFO] 
Warning:  org.apache.bookkeeper.bookie.UpgradeTest.testUpgradeCurrent
Error:    Run 1: UpgradeTest.testUpgradeCurrent:228->testUpgradeProceedure:179->Object.wait:-2 » TestTimedOut
[INFO]   Run 2: PASS
```

File: bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java [hunk 1]
```diff
@@ -1080,7 +1083,19 @@ journalFormatVersionToWrite, getBufferedChannelBuilder(),
                             .registerSuccessfulEvent(MathUtils.elapsedNanos(dequeueStartTime), TimeUnit.NANOSECONDS);
                     }
 
-                    if (numEntriesToFlush == 0) {
+                    if (localQueueEntries.isEmpty()) {
+                        queue.drainTo(localQueueEntries);
+                    }
+
+                    if (!localQueueEntries.isEmpty()) {
```
Comment (hangc0276) [inline bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java:1090]: If new entries keep put into the `queue`, the `localQueueEntries` won't be empty, it won't go into line#1107 branch, and the flush won't be triggered.

Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java:1090]: I'm not sure I got the scenario here. I think every time the loop goes back to the top on the `while (true)` and we should always check first the `localQueueEntries`  before attempting the `poll()`

Comment (merlimat) [reply bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/Journal.java]: Very good points. I actually think we should do it both before the `poll()` and the `take()`. 

@zymap @hangc0276  PTAL again

**Pull request merged**